_posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Optimize write to neo4j recommendation engine 
+author: Radoje Albijanic
+date: '2019-07-03 11:24:00 +0200'
+header: 
+  image: "images/posts/neo4j.jpg"
+  teaser: images/posts/neo4j.jpg
+categories:
+- Symfony
+- Neo4j
+- Recommendation engine
+--- 
+
+In today applications recommendations became quite necessary for better audience/customer targeting. So, as you could assume, we have our own recommendation engine for that, 
+using Symfony and Neo4j as a database. Communication between the main application and the recommendation engine is done via messages sent to/consumed from RabbitMQ. 
+We are sending some interesting events that happened on our website to the recommendation engine, and also pulling some suggestions from it.
+
+##The problem
+
+Over time we noticed something – our message queue got quite clogged, over 2 millions of messages were waiting to be written in the database! So we dug up a bit and found out 
+that Neo4j is not quite optimized for one by one writing, it is recalculating relations after each insert, and we had large amounts of messages that were just coming and coming 
+in and waiting to be processed. So we needed to fix that, 
+in order to have up to date recommendations for our system.
+
+##The investigation
+
+After reading some articles, and Neo4j official docs, we found several guidelines:
+- don’t use the ORM
+- use [UNWIND](https://neo4j.com/docs/cypher-manual/current/clauses/unwind/)   
+
+We decided to check how that really impacts the write queries to the database so we made a little experiment. We ran the insert of 500 entries with ORM, without ORM and using UNWIND:
```

Comment:
Use 5000 entries instead. It blows up the numbers more =)

Reply:
yea, I thought that and then I did not want to wait that long :D